core/src/main/scala/org/apache/spark/status/api/v1/AllJobsResource.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.status.api.v1
+
+import java.util
+import java.util.Date
+import javax.ws.rs._
+import javax.ws.rs.core.MediaType
+
+import org.apache.spark.ui.SparkUI
+import org.apache.spark.ui.jobs.JobProgressListener
+import org.apache.spark.ui.jobs.UIData.JobUIData
+
+@Produces(Array(MediaType.APPLICATION_JSON))
+private[v1] class AllJobsResource(uiRoot: UIRoot) {
+
+  @GET
+  def jobsList(
+    @PathParam("appId") appId: String,
+    @QueryParam("status") statuses: java.util.List[JobStatus]
+  ): Seq[JobData] = {
+    uiRoot.withSparkUI(appId) { ui =>
+      val statusToJobs: Seq[(JobStatus, Seq[JobUIData])] =
+        AllJobsResource.getStatusToJobs(ui)
+      val adjStatuses: util.List[JobStatus] = {
+        if (statuses.isEmpty) {
+          java.util.Arrays.asList(JobStatus.values: _*)
+        }
+        else {
+          statuses
+        }
+      }
+      val jobInfos = for {
+        (status, jobs) <- statusToJobs
+        job <- jobs if adjStatuses.contains(status)
+      } yield {
+        AllJobsResource.convertJobData(job, ui.jobProgressListener, false)
+      }
+      jobInfos.sortBy{- _.jobId}
+    }
+  }
+
+}
+
+private[v1] object AllJobsResource {
+
+  def getStatusToJobs(ui: SparkUI): Seq[(JobStatus, Seq[JobUIData])] = {
+    val statusToJobs = ui.jobProgressListener.synchronized {
+      Seq(
+        JobStatus.RUNNING -> ui.jobProgressListener.activeJobs.values.toSeq,
```

Comment:
It does make a copy, though the scala docs are a little confusing on this.  It will return a `scala.collection.Seq`, which in theory could be a `mutable.Seq`.  It actually always returns an `immutable.Seq`, however.  Even if you start with a `mutable.Seq` and call `.toSeq`, you get a copy in an `immutable.Seq`.  (If you start w/ an `immutable.Seq`, then you get back the same object, but that would be OK since its immutable in any case.)  Do you think thats OK, or should we do something more defensive here?

(You may be thinking of `Map.mapValues`, which is lazy)